### PR TITLE
Introduce system topic.

### LINKED ISSUE #409: [MIP-4] Introduce system topic to support client id unique
## Motivation
The MQTT spec calls for terminating existing clients with the same ClientID on CONNECT. But currently, MoP is only implemented on the proxy or broker side, not across clusters.

## Goal
Introduce system topic and internal reader to support this.

## Implementation

We define a system topic called `__change_events`, when MoP starts, the internal reader will try to read this topic.
Any client connects to the broker, it will send a new event(contain client id) to the system topic, and other readers try to read and check if it contains the client id. When contains, it will try to close the connection.

This can keep the cluster that only one client id is online but due to the implementation, close the same client id in another MoP instance may delay.

### PR DESCRIPTION
**Pull request by Technoboy-**
Fixes #409

Master Issue: #409

### Motivation

Introduce system topic to support cluster features.

### Documentation

- [x] `no-need-doc`

### CODE REVIEW

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/ActionType.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.streamnative.pulsar.handlers.mqtt.support.event;
+
+public enum ActionType {
+
```
Comment (mattisonchao) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/ActionType.java:18]:
```suggestion
```

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/ConnectEvent.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+@NoArgsConstructor
+@AllArgsConstructor
+public class ConnectEvent extends SourceEvent {
+
+    private String clientId;
+
+    private String ip;
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (mattisonchao) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/ConnectEvent.java:35]: Lombok @data may already override ``equals`` and ``hashcode``

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/EventType.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.streamnative.pulsar.handlers.mqtt.support.event;
+
+public enum EventType {
+
```
Comment (mattisonchao) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/EventType.java:17]:
```suggestion
```

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/SystemTopicBasedSystemEventService.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    private final ConcurrentMap<String, String> clusterClientIds;
+    private final List<EventListener> listeners;
+
+    private volatile SystemTopicClient.Reader<MqttEvent> reader;
+    private final AtomicBoolean initReader = new AtomicBoolean(false);
+    private final AtomicInteger maxRetry = new AtomicInteger(0);
+
+    public SystemTopicBasedSystemEventService(PulsarService pulsarService) {
+        this.pulsarService = pulsarService;
+        this.topicName = TopicName.get("pulsar/system/__mqtt_event");
```
Comment (mattisonchao) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/SystemTopicBasedSystemEventService.java:61]: Maybe we can extract ``pulsar/system/__mqtt_event`` to a constant. This makes it more clear that it's a system topic.

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/SystemTopicBasedSystemEventService.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    public CompletableFuture<Void> sendEvent(MqttEvent event) {
+        CompletableFuture<SystemTopicClient.Writer<MqttEvent>> writerFuture = systemTopicClient.newWriterAsync();
+        return writerFuture.thenCompose(writer -> {
+            CompletableFuture<MessageId> writeFuture = ActionType.DELETE.equals(event.getActionType())
+                    ? writer.deleteAsync(event) : writer.writeAsync(event);
+            writeFuture.whenComplete((__, ex) -> {
+                if (ex != null) {
+                    log.error("[{}] send event error.", topicName, ex);
+                }
+                writer.closeAsync();
```
Comment (mattisonchao) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/SystemTopicBasedSystemEventService.java:124]: Looks we need to wait for this method complete.

Comment (Technoboy-) [reply mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/SystemTopicBasedSystemEventService.java:123]: As we discuss, it's ok here.

**Review by mattisonchao — APPROVED** [review]

**Pull request merged**